File: CanaryServer/src/main/java/org/canary/server/service/CanaryService.java
package org.canary.server.service;

import org.apache.commons.lang3.StringUtils;
import org.apache.log4j.Logger;
import org.canary.server.model.Canary;
import org.canary.server.repository.CanaryRepository;
import org.canary.server.repository.CrudRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class CanaryService extends AbstractService<Canary> {

    @Autowired
    private CanaryRepository repository;

    private static final Logger LOGGER = Logger.getLogger(CanaryService.class);

    private CanaryService() {
	super();
    }

    @Override
    @Transactional
    @PreAuthorize("hasAuthority('USER')")
    public Canary create(final Object... args) {

	final String message = (String) args[0];

	LOGGER.debug("create[message=" + message + "]");

	if (StringUtils.isBlank(message)) {

	    throw new IllegalArgumentException(
		    "Illegal argument; message cannot be blank.");
	}

	return this.repository.create(message);
    }

    @Override
    public final CrudRepository<Canary> getRepository() {

	LOGGER.debug("getRepository");

	return this.repository;
    }

}

File: CanaryClient/src/js/service/CanaryService.js
var module = angular.module("org.canary.service");

var service = function($log, canaryRepository) {

	$log.log("CanaryService : constructor[$log=" + $log + ", canaryRepository=" + canaryRepository + "]");

	var self = this;

	self.create = function(message, callback) {

		$log.log("CanaryService : create[message=" + message + ", callback=" + (callback != null) + "]");

		return canaryRepository.create(message, callback);
	};

	self.read = function(canaryId, callback) {

		$log.log("CanaryService : read[canaryId=" + canaryId + ", callback=" + (callback != null) + "]");

		canaryRepository.read(canaryId, callback);
	};

	self.readAll = function(callback) {

		$log.log("CanaryService : readAll[callback=" + (callback != null) + "]");

		canaryRepository.readAll(callback);
	};

	self.update = function(canaryId, message, callback) {

		$log.log("CanaryService : update[canaryId=" + canaryId + ", message=" + message + ", callback=" + callback + "]");
		
		return canaryRepository.update(canaryId, message, callback);	
	};

	self.delete = function(canary, callback) {

		$log.log("CanaryService : delete[canary=" + canary + ", callback=" + callback + "]");

		return canaryRepository.delete(canary.id, callback);
	};

	return {

		create: self.create,
		read: self.read,
		readAll: self.readAll,
		update: self.update,
		delete: self.delete
	};

};

module.factory("CanaryService", ["$log", "CanaryRepository", service]);
